### PR TITLE
show test_that() calls in document outline

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

addresses #11082

### Approach

Specializes `TYPE_BRACE` for `test_that()` calls. 

Using emoji ✅ instead of having redundant `test_that()` everywhere. 

image

But using `test_that()` in the thing below: 

image

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
oh, this should not use the function icon here: 

image

**Comment by romainfrancois** [comment]
Should probably be a single test, i.e. an icon with a single ✅ instead of 3. 

image

**Comment by romainfrancois** [comment]
Separated this from #11102 because it's really two different things

**Comment by romainfrancois** [comment]
cc @Hadley @maelle you might have views on how to display them on the outline and the status bar.

File: src/gwt/acesupport/acemode/r_scope_tree.js [hunk 1]
```diff
@@ -512,6 +513,25 @@ define('mode/r_scope_tree', ["require", "exports", "module"], function(require,
...
+         debuglog("adding test_that() brace-scope " + desc);
+         var name = "test_that(" + desc + ")";
+         this.$root.addNode(
+            new this.$ScopeNodeFactory(
+               "test_that(" + desc + ")",
+               scopePos,
+               startPos,
+               ScopeNode.TYPE_BRACE,
+               {
+                  "name": "✅ " + desc,
```
Comment (kevinushey) [inline src/gwt/acesupport/acemode/r_scope_tree.js:526]: IMHO we should just use the name as-is, and consider an icon or styling (if any) on the GWT side.

(In addition, some fonts have trouble rendering certain UTF-8 symbols, so if we do want an icon or embellishment I think we should use an image)

It might also suffice to just use italics for the text label, or something similar, or leave the quotes in (since only test names would be quoted)

**Comment by maelle** [comment]
So cool to see this in progress :grin: :pray: 
I'd prefer not to read too much italic but don't have any informed opinion. :innocent:

**Comment by romainfrancois** [comment]
Maybe with some custom style, i.e. a colour: 

image

here dark green, but could be inspired from test that ? e.g. some dark red ?

**Comment by romainfrancois** [comment]
I think leaving the quotes is overwhelming, and an icon or emoji as before would be too much. We would rarely functions and tests anyway in a single test file. It's better practice to have helper functions in separate files. 

cc @jennybc you might have views here.

**Comment by jennybc** [comment]
This looks very nifty!

It's interesting that this intersects with the question of "what about code in the test file that's *not* inside `test_that()`?". As you say @romainfrancois, it's becoming more clear that that is often a smell. So it's interesting if the UI helps to discourage that practice.

What happens currently in this PR in the document outline for top-level code outside of `test_that()`? Just in general and for the specific case of a function definition.

**Comment by romainfrancois** [comment]
> What happens currently in this PR in the document outline for top-level code outside of test_that()? Just in general and for the specific case of a function definition.

Not much because the outline would only show functions and sections. but we could treat functions differently depending whether they belong or not. 

image

but maybe it's more a job for source marker annotations ?

image

Or some more ambitious dedicated "Test" pane that would enforce good practice.

**Comment by jennybc** [comment]
I think not appearing in the outline (non-function assignments) or appearing in a different colour (function definitions) is reasonable. It weakly discourages and signals that a bunch of `test_that()` calls is what one most expects to see in such a file.

**Comment by kevinushey** [comment]
For existing prior art, if users have opted into showing both chunks and sections in the document outline:

Screen Shot 2022-05-05 at 12 27 08 PM

The display is like so:

Screen Shot 2022-05-05 at 12 27 19 PM

That is, we use italics for the chunk names, to visually distinguish them from other sections.

However, it's possible (as @maelle said) that italics could become overwhelming if _every_ entry in the document outline were italicized, though. Can you share a screenshot with test labels using italics + no extra color, for comparison?

As an aside, if we choose a color for the text we need to make sure it remains visible in different editor themes (especially Solarized and dark themes).

We could also invert the relationship and choose to display R functions with italics for test files, and regular tests with "normal" styling.

**Comment by romainfrancois** [comment]
Going with altering opacity for secondary things (i.e. functions in test files). This should be more theme neutral, and nodes fade out rather than stand out (with italic): 

For example, `dplyr/test-mutate` looks like this: 

image

in solarized: 

image

**Comment by maelle** [comment]
is this enough contrast for all users?

**Comment by romainfrancois** [comment]
Oops did not mean secondary to be italic. 

image

image

It feels like a good compromise between having the item visible but more discrete. Happy to tweak the `opacity` up or down.

**Comment by kevinushey** [comment]
I like this as well. However, I agree we need to make sure the contrast remains high enough. If you're testing with a development build of RStudio Server + Chrome, you could use the built-in DevTools color contrast ratio checker:

https://juxtdesign.cc/guides/using-chrome-dev-tools-for-color-accessibility/

I think an opacity around 0.8 would probably be more appropriate.

File: src/gwt/acesupport/acemode/r_scope_tree.js [hunk 2]
```diff
@@ -92,6 +92,7 @@ define('mode/r_scope_tree', ["require", "exports", "module"], function(require,
       this.isChunk = function() { return this.scopeType == ScopeNode.TYPE_CHUNK; };
       this.isSection = function() { return this.scopeType == ScopeNode.TYPE_SECTION; };
       this.isFunction = function() { return this.isBrace() && !!this.attributes.args; };
+      this.isTest = function() { return this.isBrace() && this.attributes.type == "test"; };
```
Comment (kevinushey) [inline src/gwt/acesupport/acemode/r_scope_tree.js:95]:
```suggestion
      this.isTest = function() { return this.isBrace() && this.attributes.type === "test"; };
```

**Review by kevinushey — COMMENTED** [review]
Code otherwise LGTM; I think we can merge once we've found an opacity that prefers contrast while de-emphasizing functions visually.

**Comment by romainfrancois** [comment]
0.8:

image

0.7: 

image

I feel like .7 is fine, I don't really see fading with .8 No strong feelings either way.

**Comment by romainfrancois** [comment]
😬 I messed up with the merging, so I see a bunch of unrelated changes, but I suppose it should be fine if squashed.

**Comment by kevinushey** [comment]
Thanks! 0.7 looks good to me there as well -- 0.8 doesn't look visually distinct enough.

File: src/cpp/session/resources/themes/dreamweaver.rstheme [hunk 3]
```diff
@@ -204,7 +204,7 @@
 .xtermBold { font-weight: bold; }
 .xtermBlur { filter: blur(1px); }
 .xtermUnderline { text-decoration: underline; }
-.xtermHyperlink { text-decoration: underline; cursor: pointer; text-decoration-style: dotted; color: #75507b; }
```
Comment (kevinushey) [inline src/cpp/session/resources/themes/dreamweaver.rstheme:207]: Are these changes intentional?

Comment (romainfrancois) [reply src/cpp/session/resources/themes/dreamweaver.rstheme:207]: I think it's a merge hiccup: 

 - I merged from here from one of your review comment
 - but I also had changed locally 

so things got messy when I `git pull` locally or something like that. But these changes are in main already, from a recent pr: https://github.com/rstudio/rstudio/pull/11147

so I think it should be good to squash.
